src/mlpack/methods/ann/layer/softmin_impl.hpp

```diff
@@ -0,0 +1,61 @@
+/**
+ * @file methods/ann/layer/softmin_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of the Softmin class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_SOFTMIN_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_SOFTMIN_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "softmin.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+Softmin<InputDataType, OutputDataType>::Softmin()
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void Softmin<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    OutputType& output)
+{ 
+  InputType inputMin = arma::repmat(arma::min(input,0), input.n_rows, 1);
+  output = arma::repmat(arma::log(arma::sum(
+      arma::exp(-(input - inputMin)),0)), input.n_rows, 1);
+  output = arma::exp(-(input - inputMin) - output);
```

Comment:
Do you have a reference for this? I'm not sure where the `log` comes in. If I remember right the definition of Softmin is:

`Softmin(x) = exp(-x) / sum(exp(-x))`

Maybe I missed something?

Reply:
When I rescale the output to be in the range (0, 1), I would do:

`Softmin(x) = exp(-x - max(x)) / sum(exp(-x - max(x)))`

But I still don't see the need for `log`.

Reply:
Hey @zoq you are right with the definition of Softmin and generally we scale Softmin with `min(x)`(the scaling is with min(x) in PyTorch too) so it would be something like `Softmin(x) = exp(-x - min(x)) / sum(exp(-x - min(x)))` And for why and where the log came from so i looked at the definition of [Softmax](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/softmax_impl.hpp#L34) in mlpack, That defination is derived by taking log on both sides of Softmax so i implemented it in a similar way

1. `log(Softmin(x)) = log(exp(-x - min(x))) - log(sum(exp(-x - min(x))))`      // Taking the log on both sides.
2. `log(Softmin(x)) = (-x - min(x)) - log(sum(exp(-x - min(x))))`                   // The log gets cancelled with the exponent in the term `log(exp(-x - min(x)))`.
3. `Softmin(x) = exp((-x - min(x)) - log(sum(exp(-x - min(x)))))`                  // The only log that remains is in the second term because of the summing operation.

And I think we implement it this way to reduce the computational cost. I haven't really done any benchmarks so i am not sure.

Let me know hat do you think about this.

Reply:
Right, same problem with the `Softmax` layer, I will go over the `Softmax`, but I don't think the implementation is correct here. We should get the same output as e.g. PyTorch:

```python
import torch.nn.functional as F
import torch as th

data = th.randn(4)

data[0] = 1.43508
data[1] = 0.8377550303
data[2] = 0.008420976
data[3] = 0.0153440186
print(data)
print(F.softmin(data, dim=0))

tensor([1.4351, 0.8378, 0.0084, 0.0153])
tensor([0.0899, 0.1635, 0.3746, 0.3720])
```

which is currently not the case. So unless you see a reason can you open another PR and fix the implementation?

Reply:
We are definitely getting the same forward and backward outputs as PyTorch: 
Our implementation:-
```c++
#include <iostream>
#include <armadillo>

using namespace std;

int main()
{
  arma::colvec input("1.43508 0.8377550303 0.008420976 0.0153440186");
  
  arma::colvec inputMin = arma::repmat(arma::min(input,0),input.n_rows,1);
  arma::colvec output = arma::repmat(arma::log(arma::sum(
      arma::exp(-(input - inputMin)),0)),input.n_rows,1);
  output = arma::exp(-(input - inputMin) - output);
  output.print("output");


  arma::colvec gy = arma::ones<arma::mat>(output.n_elem);
  gy(1) = 0;
  gy(3) = 0;
  arma::colvec g = output % (gy - (arma::repmat(arma::sum(gy % output,0), output.n_rows, 1)));
  g.raw_print("g");
  return 0;
}
```
output
  ` 0.0899
   0.1635
   0.3746
   0.3720`
g
`0.0481615
-0.0759292
0.200581
-0.172813`

PyTorch:-
```python
import torch
import torch.nn as nn

data = torch.tensor([1.43508, 0.8377550303, 0.008420976, 0.0153440186], requires_grad = True) 

softmin = nn.Softmin()
out = softmin.forward(data)
print("forward" + str(out))
print("derivative: " + str(out.grad_fn(torch.tensor([1.0, 0.0, 1.0, 0.0]))))
```
`forwardtensor([0.0899, 0.1635, 0.3746, 0.3720], grad_fn=<SoftmaxBackward>)`
`derivative: tensor([ 0.0482, -0.0759,  0.2006, -0.1728], grad_fn=<SoftmaxBackwardDataBackward>)`

Let me know if i am still overlooking something.

Reply:
Just a addition: 
```python
torch.nn.functional.softmin()
```
```python
torch.nn.Softmin()
```
They both produce the same output, I think the first a function or a object and the later is a class.

Reply:
Ahh, I run my test with the expected output as input, okay, great the output is the same. So the implementation is correct, great.

Reply:
But I'm still wondering if the `log` approach is the best here especially in combination with `repmat`. `max` and `each_col` should be faster compared with creating a temporary matrix.

Reply:
We can do one thing. I will setup a testing environment with both the implementations and then I can generate some huge column vectors and run the tests multiple time for both the implementations to get a average reading. can you guide me to some tools for this. i haven't really bench-marked anything in C++

Reply:
You could use `chrono` - https://en.cppreference.com/w/cpp/chrono for the timing.

Reply:
Hey @zoq, I am back with the results you can see the testing code for the current implementation [here](https://gist.github.com/Aakash-kaushik/a51a1543ef76b8c0a8285e5ce51c5de4) and for the one you suggested [here](https://gist.github.com/Aakash-kaushik/d877015f971001787fd2e43f0b2596c9) and yes you were right the implementation you suggested is considerably fast. 
execution time for log implementation: `222150`
execution time for implementation you suggested: `149912`
Approx `48%` faster than the current implementation.

So I will go ahead and open a PR for both Softmin and Softmax. If that's fine ?

Reply:
Sounds great, thanks for looking into the different implementations.

Reply:
Hey, I have created a PR: #2657 regarding this, give it a look when you are free.